src/libc_gen/syslog.3
.\"	$NetBSD: syslog.3,v 1.24 2006/11/22 17:23:25 christos Exp $
.\"	$OpenBSD: syslog.3,v 1.25 2005/07/22 03:16:58 jaredy Exp $
.\"
.\" Copyright (c) 1985, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)syslog.3	8.1 (Berkeley) 6/4/93
.\"
.Dd August 13, 2008
.Dt SYSLOG 3
.Os
.Sh NAME
.Nm syslog ,
.Nm syslog_r ,
.Nm vsyslog ,
.Nm vsyslog_r ,
.Nm syslogp ,
.Nm syslogp_r ,
.Nm vsyslogp ,
.Nm vsyslogp_r ,
.Nm openlog ,
.Nm openlog_r ,
.Nm closelog ,
.Nm closelog_r ,
.Nm setlogmask ,
.Nm setlogmask_r
.Nd control system log
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In syslog.h
.Ft void
.Fn syslog "int priority" "const char *message" "..."
.Ft void
.Fn syslog_r "int priority" "struct syslog_data *data" "const char *message" "..."
.Ft void
.Fn syslogp "int priority" "const char *msgid" "const char *sdfmt" "const char *message" "..."
.Ft void
.Fn syslogp_r "int priority" "struct syslog_data *data" "const char *msgid" "const char *sdfmt" "const char *message" "..."
.\" .Ft void
.\" .Fn syslog_ss "int priority" "struct syslog_data *data" "const char *message" "..."
.Ft void
.Fn openlog "const char *ident" "int logopt" "int facility"
.Ft void
.Fn openlog_r "const char *ident" "int logopt" "int facility" "struct syslog_data *data"
.Ft void
.Fn closelog void
.Ft void
.Fn closelog_r "struct syslog_data *data"
.Ft int
.Fn setlogmask "int maskpri"
.Ft int
.Fn setlogmask_r "int maskpri" "struct syslog_data *data"
.In stdarg.h
.Ft void
.Fn vsyslog "int priority" "const char *message" "va_list args"
.Ft void
.Fn vsyslog_r "int priority" "struct syslog_data *data" "const char *message" "va_list args"
.Ft void
.Fn vsyslogp "int priority" "const char *msgid" "const char *sdfmt" "const char *message" "va_list args"
.Ft void
.Fn vsyslogp_r "int priority" "struct syslog_data *data" "const char *msgid" "const char *sdfmt" "const char *message" "va_list args"
.\" .Ft void
.\" .Fn vsyslog_ss "int priority" "struct syslog_data *data" "const char *message" "va_list args"
.Bd -literal

struct syslog_data {
        int             log_file;
        int             connected;
        int             opened;
        int             log_stat;
        const char     *log_tag;
        int             log_fac;
        int             log_mask;
};

#define SYSLOG_DATA_INIT { \e
    .log_file = -1, \e
    .log_fac = LOG_USER, \e
    .log_mask = 0xff, \e
}
.Ed
.Sh DESCRIPTION
The
.Fn syslog
function
writes
.Fa message
to the system message logger.
The message is then written to the system console, log files,
logged-in users, or forwarded to other machines as appropriate (See
.Xr syslogd 8 ) .
.Pp
The message is identical to a
.Xr printf 3
format string, except that
.Ql %m
is replaced by the current error
message.
(As denoted by the global variable
.Va errno ;
see
.Xr strerror 3 . )
A trailing newline is added if none is present.
.\" shouldn't the newline statement be removed?
.\" when logging through a socket a newline is
.\" not added nor is it required. -- ms
The
.Fn syslog_r
function is a multithread-safe version of the
.Fn syslog
function.
It takes a pointer to a
.Fa syslog_data
structure which is used to store
information.
This parameter must be initialized before
.Fn syslog_r
is called.
The
.Dv SYSLOG_DATA_INIT
constant is used for this purpose.
The
.Fa syslog_data
structure is composed of the following elements:
.Bl -tag -width connected
.It Dv log_file
contains the file descriptor of the file where the message is logged
.It Dv connected
indicates if connect has been done
.It Dv opened
indicates if
.Fn openlog_r
has been called
.It Dv log_stat
status bits, set by
.Fn openlog_r
.It Dv log_tag
string to tag the entry with
.It Dv log_fac
facility code
.It Dv log_mask
mask of priorities to be logged
.El
.\" .Pp
.\" The
.\" .Fn syslog_ss
.\" is the async-signal-safe version of
.\" .Fn syslog_r
.\" and is also multithread-safe.
.\" It has the following limitations:
.\" .Bl -enum -offset indent
.\" .It
.\" The format string cannot contain multi-byte character sequences.
.\" .It
.\" Floating point formats are not supported and print
.\" .Dq UNK .
.\" .It
.\" The time of the event is not sent to
.\" .Xr syslogd 8 .
.\" .It
.\" The error string in the %m format is not printed symbolically but as
.\" .Dq Error %d .
.\" .El
.\" .Pp
.\" For more information about async-signal-safe functions and signal handlers, see
.\" .Xr signal 7 .
.Pp
The
.Fn vsyslog
function
is an alternative form in which the arguments have already been captured
using the variable-length argument facilities of
.Xr varargs 3 .
.Pp
The
.Fn syslogp
variants take additional arguments which correspond to new fields in the
syslog-protocol message format. All three arguments are evaluated as
.Xr printf 3
format strings and any of them can be
.Fa NULL .
This enables applications to use message IDs, structured data, and UTF-8 encoded
content in messages.
.Pp
The message is tagged with
.Fa priority .
Priorities are encoded as a
.Fa facility
and a
.Em level .
The facility describes the part of the system
generating the message.
The level is selected from the following
.Em ordered
(high to low) list:
.Bl -tag -width LOG_AUTHPRIV
.It Dv LOG_EMERG
A panic condition.
This is normally broadcast to all users.
.It Dv LOG_ALERT
A condition that should be corrected immediately, such as a corrupted
system database.
.It Dv LOG_CRIT
Critical conditions, e.g., hard device errors.
.It Dv LOG_ERR
Errors.
.It Dv LOG_WARNING
Warning messages.
.It Dv LOG_NOTICE
Conditions that are not error conditions,
but should possibly be handled specially.
.It Dv LOG_INFO
Informational messages.
.It Dv LOG_DEBUG
Messages that contain information
normally of use only when debugging a program.
.El
.Pp
The
.Fn vsyslog_r
is used the same way as
.Fn vsyslog
except that it takes an additional pointer to a
.Fa syslog_data
structure.
It is a multithread-safe version of the
.Fn vsyslog
function described above.
.\" The
.\" .Fn vsyslog_ss
.\" is the async-signal-safe version of
.\" .Fn vsyslog_r ,
.\" is also multithread-safe, and has the same limitations as
.\" .Fn syslog_ss .
.Pp
The
.Fn openlog
function
provides for more specialized processing of the messages sent
by
.Fn syslog
and
.Fn vsyslog .
The parameter
.Fa ident
is a string that will be prepended to every message.
The
.Fa logopt
argument
is a bit field specifying logging options, which is formed by
.Tn OR Ns 'ing
one or more of the following values:
.Bl -tag -width LOG_AUTHPRIV
.It Dv LOG_CONS
If
.Fn syslog
cannot pass the message to
.Xr syslogd 8
it will attempt to write the message to the console
.Pq Dq Pa /dev/console .
.It Dv LOG_NDELAY
Open the connection to
.Xr syslogd 8
immediately.
Normally the open is delayed until the first message is logged.
Useful for programs that need to manage the order in which file
descriptors are allocated.
.It Dv LOG_PERROR
Write the message to standard error output as well to the system log.
.It Dv LOG_PID
Log the process id with each message: useful for identifying
instantiations of daemons.
(This PID is placed within brackets
between the ident and the message.)
.El
.Pp
The
.Fa facility
parameter encodes a default facility to be assigned to all messages
that do not have an explicit facility encoded:
.Bl -tag -width LOG_AUTHPRIV
.It Dv LOG_AUTH
The authorization system:
.Xr login 1 ,
.Xr su 1 ,
.Xr getty 8 ,
etc.
.It Dv LOG_AUTHPRIV
The same as
.Dv LOG_AUTH ,
but logged to a file readable only by
selected individuals.
.It Dv LOG_CRON
The cron daemon:
.Xr cron 8 .
.It Dv LOG_DAEMON
System daemons, such as
.Xr routed 8 ,
that are not provided for explicitly by other facilities.
.It Dv LOG_FTP
The file transfer protocol daemon:
.Xr ftpd 8 .
.It Dv LOG_KERN
Messages generated by the kernel.
These cannot be generated by any user processes.
.It Dv LOG_LPR
The line printer spooling system:
.Xr lpr 1 ,
.Xr lpc 8 ,
.Xr lpd 8 ,
etc.
.It Dv LOG_MAIL
The mail system.
.It Dv LOG_NEWS
The network news system.
.It Dv LOG_SYSLOG
Messages generated internally by
.Xr syslogd 8 .
.It Dv LOG_USER
Messages generated by random user processes.
This is the default facility identifier if none is specified.
.It Dv LOG_UUCP
The uucp system.
.It Dv LOG_LOCAL0
Reserved for local use.
Similarly for
.Dv LOG_LOCAL1
through
.Dv LOG_LOCAL7 .
.El
.Pp
The
.Fn openlog_r
function is the multithread-safe version of the
.Fn openlog
function.
It takes an additional pointer to a
.Fa syslog_data
structure.
This function must be used in conjunction with the other
multithread-safe functions.
.Pp
The
.Fn closelog
function
can be used to close the log file.
.Pp
The
.Fn closelog_r
does the same thing as
.Xr closelog 3
but in a multithread-safe way and takes an additional
pointer to a
.Fa syslog_data
structure.
.Pp
The
.Fn setlogmask
function
sets the log priority mask to
.Fa maskpri
and returns the previous mask.
Calls to
.Fn syslog
with a priority not set in
.Fa maskpri
are rejected.
The mask for an individual priority
.Fa pri
is calculated by the macro
.Fn LOG_MASK pri ;
the mask for all priorities up to and including
.Fa toppri
is given by the macro
.Fn LOG_UPTO toppri .
The default allows all priorities to be logged.
.Pp
The
.Fn setlogmask_r
function is the multithread-safe version of
.Fn setlogmask .
It takes an additional pointer to a
.Fa syslog_data
structure.
.Sh RETURN VALUES
The routines
.Fn closelog ,
.Fn closelog_r ,
.Fn openlog ,
.Fn openlog_r ,
.Fn syslog ,
.Fn syslog_r ,
.Fn vsyslog ,
.Fn vsyslog_r ,
.Fn syslogp ,
.Fn syslogp_r ,
.Fn vsyslogp ,
and
.Fn vsyslogp_r
return no value.
.Pp
The routines
.Fn setlogmask
and
.Fn setlogmask_r
always return the previous log mask level.
.Sh EXAMPLES
.Bd -literal -offset indent -compact
syslog(LOG_ALERT, "who: internal error 23");

openlog("ftpd", LOG_PID | LOG_NDELAY, LOG_FTP);

setlogmask(LOG_UPTO(LOG_ERR));

syslog(LOG_INFO, "Connection from host %d", CallingHost);

syslog(LOG_INFO|LOG_LOCAL2, "foobar error: %m");

syslogp(LOG_INFO|LOG_LOCAL2, NULL, NULL, "foobar error: %m");

syslogp(LOG_INFO, "ID%d", "[meta language=\\"en-US\\"]",
        "event: %s", 42, EventDescription);
.Ed
.Pp
For the multithread-safe functions:
.Bd -literal -offset indent
struct syslog_data sdata = SYSLOG_DATA_INIT;

syslog_r(LOG_INFO|LOG_LOCAL2, \*[Am]sdata, "foobar error: %m");
.Ed
.Sh SEE ALSO
.Xr logger 1 ,
.Xr syslogd 8
.Rs
.%R RFC
.%N 3164
.%D August 2001
.%T The BSD syslog Protocol
.Re
.Rs
.%R Internet-Draft
.%N draft-ietf-syslog-protocol-23
.%D September 2007
.%T The syslog Protocol
.Re
.Sh HISTORY
These non-multithread-safe functions appeared in
.Bx 4.2 .
The multithread-safe functions appeared in
.Ox 3.1
and then in
.Nx 4.0 .
The async-signal-safe functions appeared in
.Nx 4.0 .
The syslog-protocol functions appeared in
??? .
.Sh CAVEATS
It is important never to pass a string with user-supplied data as a
format without using
.Ql %s .
An attacker can put format specifiers in the string to mangle your stack,
leading to a possible security hole.
This holds true even if you have built the string
.Dq by hand
using a function like
.Fn snprintf ,
as the resulting string may still contain user-supplied conversion specifiers
for later interpolation by
.Fn syslog .
.Pp
Always be sure to use the proper secure idiom:
.Bd -literal -offset indent
syslog(priority, "%s", string);
.Ed
.Pp
With
.Fn syslogp
the caller is responsible to use the right formatting for the message fields. A
.Fa msgid
must only contain up to 32 ASCII characters. A
.Fa sdfmt
has strict rules for paranthesis and character quoting. If the
.Fa msgfmt
contains UTF-8 characters, then it has to start with a Byte Order Mark.


src/libc_gen/syslog.c
/*	$NetBSD: syslog.c,v 1.39 2006/11/22 17:23:25 christos Exp $	*/

/*
 * Copyright (c) 1983, 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if defined(LIBC_SCCS) && !defined(lint)
#if 0
static char sccsid[] = "@(#)syslog.c	8.5 (Berkeley) 4/29/95";
#else
__RCSID("$NetBSD: syslog.c,v 1.39 2006/11/22 17:23:25 christos Exp $");
#endif
#endif /* LIBC_SCCS and not lint */

#include "namespace.h"
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/syslog.h>
#include <sys/uio.h>
#include <sys/un.h>
#include <netdb.h>

#include <sys/param.h>
#include <errno.h>
#include <fcntl.h>
#include <paths.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include "reentrant.h"
#include "extern.h"

#ifdef __weak_alias
__weak_alias(closelog,_closelog)
__weak_alias(openlog,_openlog)
__weak_alias(setlogmask,_setlogmask)
__weak_alias(syslog,_syslog)
__weak_alias(vsyslog,_vsyslog)
__weak_alias(syslogp,_syslogp)
__weak_alias(vsyslogp,_vsyslogp)

__weak_alias(closelog_r,_closelog_r)
__weak_alias(openlog_r,_openlog_r)
__weak_alias(setlogmask_r,_setlogmask_r)
__weak_alias(syslog_r,_syslog_r)
__weak_alias(vsyslog_r,_vsyslog_r)
__weak_alias(syslog_ss,_syslog_ss)
__weak_alias(vsyslog_ss,_vsyslog_ss)
__weak_alias(syslogp_r,_syslogp_r)
__weak_alias(vsyslogp_r,_vsyslogp_r)
__weak_alias(syslogp_ss,_syslogp_ss)
__weak_alias(vsyslogp_ss,_vsyslogp_ss)
#endif

static struct syslog_data sdata = SYSLOG_DATA_INIT;
static void	openlog_unlocked_r(const char *, int, int,
    struct syslog_data *);
static void	disconnectlog_r(struct syslog_data *);
static void	connectlog_r(struct syslog_data *);

#define LOG_SIGNAL_SAFE	(int)0x80000000
 
#ifdef _REENTRANT
static mutex_t	syslog_mutex = MUTEX_INITIALIZER;
#endif

static char hostname[MAXHOSTNAMELEN];

/*
 * syslog, vsyslog --
 *	print message on log file; output is intended for syslogd(8).
 */
void
syslog(int pri, const char *fmt, ...)
{
        va_list ap;

        va_start(ap, fmt);
        vsyslog(pri, fmt, ap);
        va_end(ap);
}

void
vsyslog(int pri, const char *fmt, va_list ap)
{
        vsyslog_r(pri, &sdata, fmt, ap);
}

/*
 * syslogp, vsyslogp --
 *      like syslog but take additional arguments for MSGID and SD
 */
void
syslogp(int pri, const char *msgid, const char *sdfmt, const char *msgfmt, ...)
{
        va_list ap;

        va_start(ap, msgfmt);
        vsyslogp(pri, msgid, sdfmt, msgfmt, ap);
        va_end(ap);
}

void
vsyslogp(int pri, const char *msgid, const char *sdfmt, const char *msgfmt, va_list ap)
{
        vsyslogp_r(pri, &sdata, msgid, sdfmt, msgfmt, ap);
}

void
openlog(const char *ident, int logstat, int logfac)
{
	openlog_r(ident, logstat, logfac, &sdata);
}

void
closelog(void)
{
	closelog_r(&sdata);
}

/* setlogmask -- set the log mask level */
int
setlogmask(int pmask)
{
	return setlogmask_r(pmask, &sdata);
}

/* Reentrant version of syslog, i.e. syslog_r() */

void
syslog_r(int pri, struct syslog_data *data, const char *fmt, ...)
{
        va_list ap;

        va_start(ap, fmt);
        vsyslog_r(pri, data, fmt, ap);
        va_end(ap);
}

void
syslogp_r(int pri, struct syslog_data *data, const char *msgid,
        const char *sdfmt, const char *msgfmt, ...)
{
        va_list ap;

        va_start(ap, msgfmt);
        vsyslogp_r(pri, data, msgid, sdfmt, msgfmt, ap);
        va_end(ap);
}

void
syslog_ss(int pri, struct syslog_data *data, const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vsyslog_r(pri | LOG_SIGNAL_SAFE, data, fmt, ap);
	va_end(ap);
}

void
syslogp_ss(int pri, struct syslog_data *data, const char *msgid,
        const char *sdfmt, const char *msgfmt, ...)
{
        va_list ap;

        va_start(ap, msgfmt);
        vsyslogp_r(pri | LOG_SIGNAL_SAFE, data, msgid, sdfmt, msgfmt, ap);
        va_end(ap);
}

void
vsyslog_ss(int pri, struct syslog_data *data, const char *fmt, va_list ap)
{
        vsyslog_r(pri | LOG_SIGNAL_SAFE, data, fmt, ap);
}

void
vsyslogp_ss(int pri, struct syslog_data *data, const char *msgid,
        const char *sdfmt, const char *msgfmt, va_list ap)
{
        vsyslogp_r(pri | LOG_SIGNAL_SAFE, data, msgid, sdfmt, msgfmt, ap);
}


void
vsyslog_r(int pri, struct syslog_data *data, const char *fmt, va_list ap)
{
        vsyslogp_r(pri, data, NULL, NULL, fmt, ap);
}

void
vsyslogp_r(int pri, struct syslog_data *data, const char *msgid,
        const char *sdfmt, const char *msgfmt, va_list ap)
{
        size_t cnt, prlen, tries;
        char ch, *p, *t;
        struct timeval tv;
        struct tm tmnow;
        time_t now;
        int fd, saved_errno;
#define TBUF_LEN        2048
#define FMT_LEN         1024
#define MAXTRIES        10
        char *stdp = NULL;      /* pacify gcc */
        char tbuf[TBUF_LEN], fmt_cpy[FMT_LEN], fmt_cat[FMT_LEN] = "";
        size_t tbuf_left, fmt_left;
        char *fmt = fmt_cat;
        int signal_safe = pri & LOG_SIGNAL_SAFE;

        pri &= ~LOG_SIGNAL_SAFE;

#define INTERNALLOG     LOG_ERR|LOG_CONS|LOG_PERROR|LOG_PID
        /* Check for invalid bits. */
        if (pri & ~(LOG_PRIMASK|LOG_FACMASK)) {
                syslog_r(INTERNALLOG | signal_safe, data,
                    "syslog_r: unknown facility/priority: %x", pri);
                pri &= LOG_PRIMASK|LOG_FACMASK;
        }

        /* Check priority against setlogmask values. */
        if (!(LOG_MASK(LOG_PRI(pri)) & data->log_mask))
                return;

        saved_errno = errno;

        /* Set default facility if none specified. */
        if ((pri & LOG_FACMASK) == 0)
                pri |= data->log_fac;

        /* Build the message. */
        p = tbuf;  
        tbuf_left = TBUF_LEN;
        
#define DEC()                                                   \
        do {                                                    \
                if (prlen >= tbuf_left)                         \
                        prlen = tbuf_left - 1;                  \
                p += prlen;                                     \
                tbuf_left -= prlen;                             \
        } while (/*CONSTCOND*/0)

        prlen = snprintf_ss(p, tbuf_left, "<%d>1 ", pri);
        DEC();

        if (!signal_safe && (gettimeofday(&tv, NULL) != -1)) {
                /* strftime() implies tzset(), localtime_r() doesn't. */
                tzset();
                now = (time_t) tv.tv_sec;
                localtime_r(&now, &tmnow);

                prlen = strftime(p, tbuf_left, "%FT%T", &tmnow);
                DEC();
                prlen = snprintf(p, tbuf_left, ".%06ld", tv.tv_usec);
                DEC();
                prlen = strftime(p, tbuf_left-1, "%z", &tmnow);
                /* strftime gives eg. "+0200", but we need "+02:00" */
                if (prlen == 5) {
                        p[prlen+1] = p[prlen];
                        p[prlen]   = p[prlen-1];
                        p[prlen-1] = p[prlen-2];
                        p[prlen-2] = ':';
                        prlen += 1;
                }
        } else {
                prlen = snprintf_ss(p, tbuf_left, "-");

                /* if gmtime_r() was signal-safe we could output the UTC-time:            
                gmtime_r(&now, &tmnow);
                prlen = strftime(p, tbuf_left, "%FT%TZ", &tmnow);
                */
        }
        DEC();
        prlen = snprintf_ss(p, tbuf_left, " %s ", hostname);
        DEC();

        if (data->log_stat & LOG_PERROR)
                stdp = p;
        if (data->log_tag == NULL)
                data->log_tag = getprogname();

        prlen = snprintf_ss(p, tbuf_left, "%s ",
            data->log_tag ? data->log_tag : "-");
        DEC();

        if (data->log_stat & LOG_PID)
                prlen = snprintf_ss(p, tbuf_left, "%d ", getpid());
        else
                prlen = snprintf_ss(p, tbuf_left, "- ");
        DEC();

        /* 
         * concat the format strings, then use one vsnprintf()
         */
        if (msgid != NULL && *msgid != '\0') {
                strlcat(fmt_cat, msgid, FMT_LEN);
                strlcat(fmt_cat, " ", FMT_LEN);
        } else
                strlcat(fmt_cat, "- ", FMT_LEN);

        if (sdfmt != NULL && *sdfmt != '\0') {
                strlcat(fmt_cat, sdfmt, FMT_LEN);
        } else
                strlcat(fmt_cat, "-", FMT_LEN);

        if (msgfmt != NULL && *msgfmt != '\0') {
                strlcat(fmt_cat, " ", FMT_LEN);
                strlcat(fmt_cat, msgfmt, FMT_LEN);
        }

        /* 
         * We wouldn't need this mess if printf handled %m, or if 
         * strerror() had been invented before syslog().
         */
        for (t = fmt_cpy, fmt_left = FMT_LEN; (ch = *fmt) != '\0'; ++fmt) {
                if (ch == '%' && fmt[1] == 'm') {
                        char ebuf[128];
                        ++fmt;
                        if (signal_safe ||
                            strerror_r(saved_errno, ebuf, sizeof(ebuf)))
                                prlen = snprintf_ss(t, fmt_left, "Error %d", 
                                    saved_errno);
                        else
                                prlen = snprintf_ss(t, fmt_left, "%s", ebuf);
                        if (prlen >= fmt_left)
                                prlen = fmt_left - 1;
                        t += prlen;
                        fmt_left -= prlen;
                } else if (ch == '%' && fmt[1] == '%' && fmt_left > 2) {
                        *t++ = '%';
                        *t++ = '%';
                        fmt++;
                        fmt_left -= 2;
                } else {
                        if (fmt_left > 1) {
                                *t++ = ch;
                                fmt_left--;
                        }
                }
        }
        *t = '\0';
        
        if (signal_safe)
                prlen = vsnprintf_ss(p, tbuf_left, fmt_cpy, ap);
        else
                prlen = vsnprintf(p, tbuf_left, fmt_cpy, ap);
        DEC();
        cnt = p - tbuf;

        /* Output to stderr if requested. */
        if (data->log_stat & LOG_PERROR) {
                struct iovec iov[2];

                iov[0].iov_base = stdp;
                iov[0].iov_len = cnt - (stdp - tbuf);
                iov[1].iov_base = __UNCONST("\n");
                iov[1].iov_len = 1;
                (void)writev(STDERR_FILENO, iov, 2);
        }

        /* Get connected, output the message to the local logger. */
        if (data == &sdata)
                mutex_lock(&syslog_mutex);
        if (!data->opened)
                openlog_unlocked_r(data->log_tag, data->log_stat, 0, data);
        connectlog_r(data);

        /*
         * If the send() failed, there are two likely scenarios:
         *  1) syslogd was restarted
         *  2) /dev/log is out of socket buffer space
         * We attempt to reconnect to /dev/log to take care of
         * case #1 and keep send()ing data to cover case #2
         * to give syslogd a chance to empty its socket buffer.
         */
        for (tries = 0; tries < MAXTRIES; tries++) {
                if (send(data->log_file, tbuf, cnt, 0) != -1)
                        break;
                if (errno != ENOBUFS) {
                        disconnectlog_r(data);
                        connectlog_r(data);
                } else
                        (void)usleep(1);
        }

        /*
         * Output the message to the console; try not to block
         * as a blocking console should not stop other processes.
         * Make sure the error reported is the one from the syslogd failure.
         */
        if (tries == MAXTRIES && (data->log_stat & LOG_CONS) &&
            (fd = open(_PATH_CONSOLE, O_WRONLY|O_NONBLOCK, 0)) >= 0 &&
            (p = strchr(tbuf, '>')) != NULL) {
                struct iovec iov[2];
                iov[0].iov_base = ++p;
                iov[0].iov_len = cnt - (p - tbuf);
                iov[1].iov_base = __UNCONST("\r\n");
                iov[1].iov_len = 2;
                (void)writev(fd, iov, 2);
                (void)close(fd);
        }

        if (data == &sdata)
                mutex_unlock(&syslog_mutex);

        if (data != &sdata)
                closelog_r(data);
}

static void
disconnectlog_r(struct syslog_data *data)
{
	/*
	 * If the user closed the FD and opened another in the same slot,
	 * that's their problem.  They should close it before calling on
	 * system services.
	 */
	if (data->log_file != -1) {
		(void)close(data->log_file);
		data->log_file = -1;
	}
	data->connected = 0;		/* retry connect */
}

static void
connectlog_r(struct syslog_data *data)
{
	/* AF_UNIX address of local logger */
	static const struct sockaddr_un sun = {
		.sun_family = AF_LOCAL,
		.sun_len = sizeof(sun),
		.sun_path = _PATH_LOG,
	};

	if (data->log_file == -1 || fcntl(data->log_file, F_GETFL, 0) == -1) {
		if ((data->log_file = socket(AF_UNIX, SOCK_DGRAM, 0)) == -1)
			return;
		(void)fcntl(data->log_file, F_SETFD, FD_CLOEXEC);
		data->connected = 0;
	}
	if (!data->connected) {
		if (connect(data->log_file,
		    (const struct sockaddr *)(const void *)&sun,
		    sizeof(sun)) == -1) {
			(void)close(data->log_file);
			data->log_file = -1;
 		} else
			data->connected = 1;
	}
}

static void
openlog_unlocked_r(const char *ident, int logstat, int logfac,
    struct syslog_data *data)
{
	if (ident != NULL)
		data->log_tag = ident;
	data->log_stat = logstat;
	if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0)
		data->log_fac = logfac;

	if (data->log_stat & LOG_NDELAY)	/* open immediately */
		connectlog_r(data);

        if (gethostname(hostname, sizeof(hostname)) == -1
                        || hostname[0] == '\0') {
                /* can this really happen? */
                hostname[0] = '-';
                hostname[1] = '\0';
                return;
        }
        if (!strchr(hostname, '.')) { /* no FQDN */
                struct addrinfo *res;
                struct addrinfo hints = {
                        .ai_family = PF_UNSPEC,
                        .ai_socktype = 0,
                        .ai_protocol = 0,
                        .ai_flags = AI_CANONNAME,
                };
                if (getaddrinfo(hostname, NULL, &hints, &res))
                        return;
                /* try to resolve back to hostname */
                (void)getnameinfo(res->ai_addr, (socklen_t)res->ai_addr->sa_len,
                        hostname, sizeof(hostname), NULL, 0, 0);
                freeaddrinfo(res);
        }
}

void
openlog_r(const char *ident, int logstat, int logfac, struct syslog_data *data)
{
	if (data == &sdata)
		mutex_lock(&syslog_mutex);
	openlog_unlocked_r(ident, logstat, logfac, data);
	if (data == &sdata)
		mutex_unlock(&syslog_mutex);
}

void
closelog_r(struct syslog_data *data)
{
	if (data == &sdata)
		mutex_lock(&syslog_mutex);
	(void)close(data->log_file);
	data->log_file = -1;
	data->connected = 0;
	data->log_tag = NULL;
	if (data == &sdata)
		mutex_unlock(&syslog_mutex);
}

int
setlogmask_r(int pmask, struct syslog_data *data)
{
	int omask;

	omask = data->log_mask;
	if (pmask != 0)
		data->log_mask = pmask;
	return omask;
}
